enterprise/server/usage/usage.go

```diff
@@ -83,41 +128,235 @@ func (ut *tracker) Increment(ctx context.Context, uc *tables.UsageCounts) error
+			return err
+		}
+	}
+	return nil
+}
+
+func (ut *tracker) flushCounts(ctx context.Context, groupID string, c collectionPeriod, counts *tables.UsageCounts) error {
+	pk := &tables.Usage{
+		GroupID:         groupID,
+		PeriodStartUsec: timeutil.ToUsec(c.UsagePeriod().Start()),
+	}
+	dbh := ut.env.GetDBHandle()
+	return dbh.Transaction(ctx, func(tx *db.DB) error {
+		// Create a row for the corresponding usage period if one doesn't already
+		// exist
+		err := tx.Clauses(clause.OnConflict{DoNothing: true}).Create(pk).Error
+		if err != nil {
+			return err
+		}
+		updates, err := usageUpdates(c, counts)
+		if err != nil {
+			return err
+		}
+		// Update the usage row, but only if collection period data has not already
+		// been written (for example, if the previous flush failed to delete the
+		// data from Redis).
+		return tx.Model(&tables.Usage{}).Where(
+			`group_id = ?
+			AND period_start_usec = ?
+			AND final_before_usec <= ?`,
+			pk.GroupID,
+			pk.PeriodStartUsec,
+			timeutil.ToUsec(c.Start()),
+		).Updates(updates).Error
+	})
+}
+
+func (ut *tracker) currentCollectionPeriod() collectionPeriod {
+	return collectionPeriodStartingAt(ut.clock.Now())
+}
+
+func (ut *tracker) oldestWritableCollectionPeriod() collectionPeriod {
+	return collectionPeriodStartingAt(ut.clock.Now().Add(-redisKeyTTL))
+}
+
+func (ut *tracker) lastSettledCollectionPeriod() collectionPeriod {
+	return collectionPeriodStartingAt(ut.clock.Now().Add(-(collectionPeriodDuration + collectionPeriodSettlingTime)))
+}
+
+// isSettled returns whether the given collection period will no longer have
+// usage data written to it and is therefore safe to flush to the DB.
+func (ut *tracker) isSettled(c collectionPeriod) bool {
+	return !time.Time(c).After(time.Time(ut.lastSettledCollectionPeriod()))
+}
+
+// collectionPeriod is an interval of time starting at the beginning of a minute
+// in UTC time and lasting one minute. Usage data is bucketed by collection
+// period in Redis.
+type collectionPeriod time.Time
+
+func collectionPeriodStartingAt(t time.Time) collectionPeriod {
 	utc := t.UTC()
-	// Start of minute
-	return time.Date(
+	return collectionPeriod(time.Date(
 		utc.Year(), utc.Month(), utc.Day(),
 		utc.Hour(), utc.Minute(), 0, 0,
-		utc.Location())
+		utc.Location()))
+}
+
+func parseCollectionPeriod(s string) (collectionPeriod, error) {
+	usec, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return collectionPeriodZeroValue, err
+	}
+	t := timeutil.FromUsec(usec)
+	return collectionPeriodStartingAt(t), nil
+}
+
+func (c collectionPeriod) Start() time.Time {
+	return time.Time(c)
+}
+
+func (c collectionPeriod) End() time.Time {
+	return c.Start().Add(collectionPeriodDuration)
+}
+
+// Next returns the next collection period after this one.
+func (c collectionPeriod) Next() collectionPeriod {
+	return collectionPeriod(c.End())
+}
+
+// UsagePeriod returns the usage period that this collection period is contained
```

Comment:
I don't totally follow the difference between a collectionPeriod and a usageperiod -- can you explain it?

Reply:
Done (added the explanation to the comment here)